src/main/kotlin/org/elm/ide/refactoring/ElmIntroduceVariableHandler.kt

```diff
@@ -0,0 +1,188 @@
+import com.intellij.util.DocumentUtil
+import org.elm.ide.utils.findExpressionAtCaret
+import org.elm.ide.utils.findExpressionInRange
+import org.elm.lang.core.psi.*
+import org.elm.lang.core.psi.elements.ElmAnonymousFunctionExpr
+import org.elm.lang.core.psi.elements.ElmCaseOfBranch
+import org.elm.lang.core.psi.elements.ElmLetInExpr
+import org.elm.lang.core.psi.elements.ElmValueDeclaration
+import org.elm.openapiext.runWriteCommandAction
+
+class ElmIntroduceVariableHandler : RefactoringActionHandler {
+
+    override fun invoke(project: Project, editor: Editor, file: PsiFile, dataContext: DataContext?) {
+        if (file !is ElmFile) return
+        val exprs = findCandidateExpressions(editor, file)
+        when (exprs.size) {
+            0 -> {
+                val message = RefactoringBundle.message(if (editor.selectionModel.hasSelection())
+                    "selected.block.should.represent.an.expression"
+                else
+                    "refactoring.introduce.selection.error"
+                )
+                val title = RefactoringBundle.message("introduce.variable.title")
+                val helpId = "refactoring.extractVariable"
+                CommonRefactoringUtil.showErrorHint(project, editor, message, title, helpId)
+            }
+            1 -> introduceVariable(editor, exprs.single())
+            else -> showExpressionChooser(editor, exprs) { chosenExpr ->
+                introduceVariable(editor, chosenExpr)
+            }
+        }
+    }
+
+    private fun findCandidateExpressions(editor: Editor, file: ElmFile): List<ElmExpressionTag> {
+        val selection = editor.selectionModel
+        return if (selection.hasSelection()) {
+            // If the user has some text selected, make a single suggestion based on the selection
+            listOfNotNull(findExpressionInRange(file, selection.selectionStart, selection.selectionEnd))
+        } else {
+            // Suggest nested expressions at caret position
+            val expr = findExpressionAtCaret(file, editor.caretModel.offset) ?: return emptyList()
+            expr.ancestors
+                    .takeWhile { it !is ElmValueDeclaration && it !is ElmLetInExpr }
+                    .filterIsInstance<ElmExpressionTag>()
+                    .toList()
+        }
+    }
+
+    private fun introduceVariable(editor: Editor, chosenExpr: ElmExpressionTag) {
+        if (!chosenExpr.isValid) return
+        val project = editor.project ?: return
+
+        val replacer = ExpressionReplacer(project, editor, chosenExpr)
+        val anchor = findAnchor(chosenExpr) ?: error("could not find a place to introduce variable")
+        when {
+            anchor is ElmLetInExpr -> replacer.extendExistingLet(anchor)
+            else -> replacer.introduceLet(anchor)
+        }
+    }
+
+    private fun findAnchor(chosenExpr: ElmExpressionTag): ElmPsiElement? {
+        // find the nearest location where a let/in would be
```

Comment:
In the future we may want to allow the user to choose which scope they would like to add it to instead of assuming that it's always the nearest scope.